src/metric/perf_counter.cpp

```diff
@@ -85,5 +99,109 @@ double PerfCounter::read()
     accumulated_ = value;
     return accumulated_;
 }
+
+CounterBuffer::CounterBuffer(std::size_t ncounters)
+: buf_(new char[2 * total_buf_size(ncounters)]), accumulated_(ncounters, 0)
+{
+    // `buf_` points to contiguous memory that holds both entries of the double
+    // buffer;  let the non-owning pointers `current_` and `previous_` point to
+    // the correct offsets in the former.
+    current_ = reinterpret_cast<ReadFormat*>(buf_.get());
+    current_->nr = ncounters;
+    current_->time_enabled = 0;
+    current_->time_running = 0;
+    std::memset(&current_->values, 0, ncounters * sizeof(uint64_t));
+
+    previous_ = reinterpret_cast<ReadFormat*>(buf_.get() + total_buf_size(ncounters));
+    previous_->nr = ncounters;
+    previous_->time_enabled = 0;
+    previous_->time_running = 0;
+    std::memset(&current_->values, 0, ncounters * sizeof(uint64_t));
+}
+
+void CounterBuffer::read(int group_leader_fd)
+{
+    auto ncounters = previous_->nr;
+    assert(accumulated_.size() == ncounters);
+
+    auto bufsz = total_buf_size(ncounters);
+
+    auto read_bytes = ::read(group_leader_fd, current_, bufsz);
+    if (read_bytes < 0 || static_cast<decltype(bufsz)>(read_bytes) != bufsz)
+    {
+        Log::error() << "failed to read counter buffer from file descriptor";
+        throw_errno();
+    }
+
+    update_buffers();
+}
+
+void CounterBuffer::read(const ReadFormat* inbuf)
+{
+    assert(accumulated_.size() == inbuf->nr);
+
+    auto ret = std::memcpy(current_, inbuf, total_buf_size(inbuf->nr));
```

Comment:
As per the standard `std::memcpy` always returns `dest` - there's no use in checking. When it fails, it segfaults.